### ISSUE TITLE
Setting 'statuscode' throws exception

### ISSUE THREAD

**Issue by Part-A** [AlexaCRM/php-crm-toolkit#24]
In attempting to set 'statuscode' as 'Complete' on a task entity, an exception is thrown:

>   [Exception]                                
>   Unable to retrieve data from Dynamics CRM  
> 

In my example, the following code is used for creating the task entity:

```
private function addTask( $entity, $data, $entityType, $taskDetail, $taskFields = [] ) {

	$id                      = $entity->ID;
	$task                    = $this->dynamics->getEntity( 'task' );
	$task->subject           = $this->getEmailSubject( $data );
	$task->regardingobjectid = $entity;
	$task->scheduledstart    = $data->timestamp;

	$task->statuscode = 5; // This line cause the issue 

	//print_r( $task->attributes['statuscode'] );
	$taskDescription = $taskDetail . "\n" . $this->getEmailSubject( $data ) . "\n";
	foreach ( $taskFields as $taskField ) {
		$taskDescription .= $taskField . ": " . $data->$taskField . "\n";
	}
	$task->description = $taskDescription;

	return $task->create();
}
```

If the line:
`	$task->statuscode = 5; // This line cause the issue  `
is removed, the task entity is created without issue.

Changing the line to:
`	$task->statuscode = 'Complete'; `
continues to cause the same issue

The attribute details specify that the option is valid for update:

```
AlexaCRM\CRMToolkit\Entity\Attribute Object
(
    [logicalName] => statuscode
    [label] => Status Reason
    [description] => Select the task's status.
    [format] => 
    [maxLength] => 0
    [isCustom] => 
    [imeMode] => 
    [isPrimaryId] => 
    [isPrimaryName] => 
    [type] => Status
    [isLookup] => 
    [isValidForCreate] => 1
    [isValidForUpdate] => 1
    [isValidForRead] => 1
    [requiredLevel] => None
    [attributeOf] => 
    [optionSet] => AlexaCRM\CRMToolkit\Entity\OptionSet Object
        (
            [name] => task_statuscode
            [type] => Status
            [displayName] => Reason for the status of the task.
            [description] => Reason for the status of the task.
            [options] => Array
                (
                    [2] => Not Started
                    [3] => In Progress
                    [4] => Waiting on someone else
                    [5] => Completed
                    [6] => Canceled
                    [7] => Deferred
                )

            [isGlobal] => 
        )

    [lookupTypes] => 
)
```

**Comment by georged** [comment]
Hi @Part-A,
this statuscode is invalid for the active state code. Try setting statecode as well:
```
$task->statecode = 1;
$task->statuscode = 5;
```

Cheers
George

**Comment by Part-A** [comment]
Unfortunately, statecode is not valid for create or update, as demonstrated here:

```AlexaCRM\CRMToolkit\Entity\Attribute Object
(
    [logicalName] => statecode
    [label] => Activity Status
    [description] => Shows whether the task is open, completed, or canceled. Completed and canceled tasks are read-only and can't be edited.
    [format] => 
    [maxLength] => 0
    [isCustom] => 
    [imeMode] => 
    [isPrimaryId] => 
    [isPrimaryName] => 
    [type] => State
    [isLookup] => 
    [isValidForCreate] => 
    [isValidForUpdate] => 
    [isValidForRead] => 1
    [requiredLevel] => SystemRequired
    [attributeOf] => 
    [optionSet] => AlexaCRM\CRMToolkit\Entity\OptionSet Object
        (
            [name] => task_statecode
            [type] => State
            [displayName] => Status of the task.
            [description] => Status of the task.
            [options] => Array
                (
                    [0] => Open
                    [1] => Completed
                    [2] => Canceled
                )

            [isGlobal] => 
        )

    [lookupTypes] => 
)
```

Therefore AlexaCRM\CRMToolkit\Entity rejects the setting of statecode

**Comment by georged** [comment]
I just tried it with SDK, it's not possible to create completed tasks. Instead, you should do something like
```
$tid = $task->create();
$task = new Entity('task', $tid);
$task->statecode = new OptionSetValue(1); // completed 
$task->statuscode = new OptionSetValue(5); // completed
$task->update();
return $tid;
```

**Comment by Part-A** [comment]
Thank you for your suggestion, your help is very much appreciated, unfortunately, it does not work.

Because 'statecode' is not valid for update as shown in the Attribute details above, this code in Entity.php blocks the update:

```
            /* Check that this property can be set in Creation or Update */
            if ( $this->attributes[ $property ]->isValidForCreate == false && $this->attributes[ $property ]->isValidForUpdate == false ) {
                trigger_error( 'Property ' . $property . ' of the ' . $this->entityLogicalName
                               . ' entity cannot be set', E_USER_NOTICE );

                return;
            }
```

Therefore, this leaves the task with the following two entries, which as you have already stated is an invalid state for 'statuscode' when 'statecode' is set as Open 

 ```           [statecode] => Array
                (
                    [Value] => AlexaCRM\CRMToolkit\OptionSetValue Object
                        (
                            [value] => 0
                            [label] => Open
                        )

                    [Changed] => 
                )
```


 ```                       [statuscode] => Array
                (
                    [Value] => AlexaCRM\CRMToolkit\OptionSetValue Object
                        (
                            [value] => 5
                            [label] => Completed
                        )

                    [Changed] => 1
                )

```

This state throws an exception on update.

**Comment by Part-A** [comment]
Temporarily disabling the code in Entity.php, does allow the setting of 'statecode'. Giving the following for the task entity:

```
            [statecode] => Array
                (
                    [Value] => AlexaCRM\CRMToolkit\OptionSetValue Object
                        (
                            [value] => 1
                            [label] => Completed
                        )

                    [Changed] => 1
                )
```

However, on update, this still throws an exception:

>   [Exception]                                
>   Unable to retrieve data from Dynamics CRM  
>

**Comment by georged** [comment]
Hmmm, statecode should be valid for update and the sample update above should work.
We'll investigate further.
Thanks
George

**Comment by georged** [comment]
@Part-A
What's your CRM version?

**Comment by Part-A** [comment]
Microsoft Dynamics® CRM 2015
(7.0.1.129) (DB 7.0.1.129)

**Comment by georged** [comment]
@Part-A, 

that would explain it. You can't update statuscode and statecode in that version, you need to use SetStateRequest. See [the docs](https://msdn.microsoft.com/en-us/library/dn932124(v=crm.7).aspx):

> For Microsoft Dynamics CRM Online organizations, this feature is available only if your organization has updated to Dynamics CRM Online 2015 Update 1. This feature is not available for Dynamics CRM (on-premises).

We've tested updating on Dynamics 365 Online instance and it worked just fine.
